src/memory_client.rs

```diff
@@ -60,20 +60,18 @@ impl<'a> MemoryClient<'a> {
         // constraints can be applied.
         //
         // In this case, we should expect `Infallible` error.
-        let state = self
-            .transactor
-            .result()
-            .expect("MemoryStorage implements `Infallible` as storage error. This means panic should be unreachable.");
-
-        if state.should_revert() {
-            self.transactor.as_mut().revert();
+        if let Ok(state) = self.transactor.result() {
+            if state.should_revert() {
+                self.transactor.as_mut().revert();
+            } else {
+                self.transactor.as_mut().commit();
+            }
```

Comment:
The memory storage implements infallible as error; its pointless to check the outcome of the result. Also, isn't this outside the scope of the problem?

Reply:
The problem is that this was panicking if transaction validation failed during init. I needed to capture the error instead of expecting here in order to capture the maturity error.

Reply:
Hmm that's very odd. If we panicked with infallible error, then the cause might be deeper. We can quick fix this and I'll investigate later.